Raft: send commit ack back to follower before closing connection

runshenzhu opened the pull request:
if there are only 2 nodes in the cluster, and leader is leaving before closing the connection, leader has to ensure that follower gets noticed about this raft conf change commit. 

Otherwise, follower would assume there are still 2 nodes in the cluster and won't get elected into the leader by acquiring the majority (2 nodes). Check out [my repo](https://github.com/runshenzhu/swarmkit/commits/raft-leave-join) to reproduce this bug.

This PR fix the problem by sending commit ack back to follower manually, before leader closing connection.

An alternative approach is to keep leader from leaving when there are only 2 nodes in the cluster.

Signed-off-by: Runshen Zhu <runshen.zhu@gmail.com>

ping: @dperny @aaronlehmann @abronan 
fix #1178

abronan commented:
Fix #1178

aaronlehmann commented on manager/state/raft/raft.go:
```diff
@@ -1206,10 +1206,63 @@ func (n *Node) applyRemoveNode(cc raftpb.ConfChange) (err error) {
+
+			conn = member
+			break
+		}
```
I think we should check that `conn` is not nil here.

runshenzhu replied:
In this case, what kind of action I should apply, return an error or continue the removal?

aaronlehmann replied:
I would log a warning and continue with the removal.

aaronlehmann commented on manager/state/raft/raft.go:
```diff
@@ -1206,10 +1206,63 @@ func (n *Node) applyRemoveNode(cc raftpb.ConfChange) (err error) {
+
+		// send with best-effort strategy
+		ctx := context.Background()
+	Loop:
```
lowercase

aaronlehmann commented:
I like this solution but I know @abronan had some concerns. @abronan WDYT?

aaronlehmann commented on manager/state/raft/raft.go:
```diff
@@ -1206,10 +1206,63 @@ func (n *Node) applyRemoveNode(cc raftpb.ConfChange) (err error) {
+		// send with best-effort strategy
+		ctx := context.Background()
+	Loop:
+		for {
```
I don't think it makes sense to have a loop here. grpc will automatically retry if it can't reach the remote host (it has a weird asynchronous `Dial` function). If we want to allow a few retries, we should just use a longer duration than `sendTimeout` - perhaps `sendTimeout * 10` or something like that. But I think we absolutely want to put a bound on the number of retries.

aaronlehmann commented on manager/state/raft/raft.go:
```diff
@@ -1206,10 +1206,75 @@ func (n *Node) applyRemoveNode(cc raftpb.ConfChange) (err error) {
+		// send with best-effort strategy
+		ctx := context.Background()
+	loop:
+		for {
```
Comment was lost because the diff changed:

I don't think it makes sense to have a loop here. grpc will automatically retry if it can't reach the remote host (it has a weird asynchronous `Dial` function). If we want to allow a few retries, we should just use a longer duration than `sendTimeout` - perhaps `sendTimeout * 10` or something like that. But I think we absolutely want to put a bound on the number of retries.

runshenzhu replied:
Yes it makes sense. I will change and re-test it.

abronan commented:
@aaronlehmann @runshenzhu It's not a problem inherent to the PR, I also think the solution is good to fix the problem.

I think there is a problem though in allowing the Leader to be demoted. The problem is even bigger in the case of 2 managers, what if the message disappears mysteriously in the network or we can't contact the remaining node during that brief period of time: the problem remains the same.

I would really prefer something like a `docker node stepdown node01` to step down as a leader and then remove the node gracefully. Here we demote but it is as safe as saying `docker node leave -f node01` as there is no guarantee that the operation will complete successfully and that the first node will pick up the leader status. Thus the user will be forced to recover the cluster using `--force-new-cluster` on the remaining node.

I'm just worried we encourage to do the wrong things for a matter of being flexible at handling the raft cluster.

Otherwise LGTM but I would encourage opening up a discussion on this operation safety-net.

aaronlehmann commented:
> I think there is a problem though in allowing the Leader to be demoted. The problem is even bigger in the case of 2 managers, what if the message disappears mysteriously in the network or we can't contact the remaining node during that brief period of time: the problem remains the same.

I'm confused, isn't the two manager case exactly what this PR solves?

It's true it won't be able to do anything if we can't contact the remaining node, but it seems unlikely we'd be able to commit the configuration change if that node is really unreachable.

dperny commented:
The two generals problem, in real life!

runshenzhu commented:
I will bring back the block mechanism and add back off re-try.

aaronlehmann commented on manager/state/raft/raft.go:
```diff
@@ -1206,10 +1206,82 @@ func (n *Node) applyRemoveNode(cc raftpb.ConfChange) (err error) {
+		}
+
+		if conn == nil {
+			n.Config.Logger.Warningf("when sending leaving commit ack to follower %x, grpc connection is lost", peerID)
```
The error condition is actually that the node was not found in the member list.

runshenzhu commented:
updated. PTAL @aaronlehmann @abronan

aaronlehmann commented on manager/state/raft/raft.go:
```diff
@@ -1206,10 +1206,82 @@ func (n *Node) applyRemoveNode(cc raftpb.ConfChange) (err error) {
+
+		ctx := context.Background()
+		backoff := 0
+		sleepTime := 10 * time.Millisecond
```
Can you change this to 100 ms? 10 ms seems a bit eager for something like this.

I'm also not sure whether we need this exponential backoff. Are there any situations where the TCP connection would succeed by the RPC would fail, and we still want to retry? If not, perhaps we can rely on GRPC's built-in exponential backoff.

abronan replied:
Agreed with aaron, would be best if we relied on gRPC's backoff mechanism instead :)

runshenzhu commented:
update: backoff is removed.

aaronlehmann commented:
I just thought of an alternative approach that might be a little more elegant.

I think what is happening is that when the removal is committed, we first call `n.send`. This starts a goroutine that sends a message to the remaining node, in this case telling it that the removal has been committed. We also call `RemoveMember`, which closes the connection. If `RemoveMember` runs before the `sendToMember` goroutine, it will close the connection before the message makes it to the remaining node.

If this is correct, I think we could wait for the sends to finish before calling `RemoveMember`. Perhaps something like this:

```go
if cc.NodeID == n.Config.ID && n.IsLeader() && len(members) == 2 { // or maybe even remove this check
        n.asyncTasks.Wait()
}

return n.cluster.RemoveMember(cc.NodeID)
```

Again, this is all theoretical, but if it's correct, it would mean less code and it would avoid possibly duplicating the message we send.

runshenzhu commented:
Yes, my original best-effort approach is just like this one. But then I though this approach is too optimistic that it never retries when getting errors, such as `time limit exceeds`

However, I do agree that `n.asyncTasks.Wait()` approach is much more elegant than the current one. I could change this PR if people like `n.asyncTasks.Wait()`.

aaronlehmann commented:
I don't have very strong feelings, but I like reducing the amount of code. So I think I'd prefer `n.asyncTasks.Wait()`.

abronan commented:
I'd prefer `n.asyncTasks.Wait()` better as well. Provided that this passes the stress test.

runshenzhu commented:
updated: use `n.asyncTasks.Wait()`

aaronlehmann commented:
LGTM if it passes your tests

runshenzhu commented:
@aaronlehmann When running my test on this approach, it gets error that's related to `time limit exceeds` in very few cases.

Since we decide to use the best effort approach, I think it's OK to have `time limit exceeds` error.

aaronlehmann commented:
Interesting, is that the stress test or a different test? Is it `sendToMember` that gets this error or something else in the test?

I'd like to understand why the timeout gets hit. I expected that the `n.asyncTasks.Wait()` would work reliably with the stress test, since it should always be possible to reach the remaining member at that point.

runshenzhu commented:
@aaronlehmann That might be a bug of my test. Here is the code:
```go
// while this Leave request might timeout, 
// the underlying Leader node is still executing the leaving and waiting on the asyncTasks, which is good.
raftClient.Leave(ctx, &api.LeaveRequest{Node: &api.RaftMember{RaftID: node.Config.ID}})

// ignore the error of leave and teardown node, which cancels the ctx of node
// and causes the last msg sending fail, which is bad
raftutils.TeardownCluster(t, map[uint64]*raftutils.TestNode{id: nodes[id]})
``` 

As the code shows, node is shut down without checking the return value of `Leave`, which makes test fail.

I don't know whether this situation could happen in real environment. If `docker swarm leave --force` shutdowns the underlying node anyway, without checking any errors. This might be a problem.

aaronlehmann commented:
I don't understand why `Leave` would time out in the test though.

runshenzhu commented:
a context with time out is passed to `raftClient.Leave` as the parameter. That's why `raftClient.Leave` times out. 

After changing the context to long running type, I didn't see failure after running stress test many times.

aaronlehmann commented:
I see. I'm still surprised this is an issue, because it looks like the timeout on the context for `Leave` was 10 seconds in your test. `sendToMember` has its own timeout, `sendTimeout`, which should be less than that. I wonder what is taking 10 seocnds.

runshenzhu commented:
context uses real clock to trigger time out, but Raft in the test uses fake clock to tick. Unlike `waitforcluster`, in which fake clock is sync-ed with real clock, these 2 clocks didn't sync with each other in `raftClient.Leave`. 

I think that's why `raftClient.Leave` times out.

aaronlehmann commented:
Sorry, I still feel like I'm missing something.

If the issue is that `raftClient.Leave` waits 10 seconds (real time) but times out because the fake clock isn't ticking, I don't understand why passing a context without a timeout to `Leave` would help. The fake clock wouldn't be advanced until `Leave` returns, so wouldn't we just have a deadlock if `Leave` is waiting for the fake clock to tick?

abronan commented:
By the way, I don't think `n.asyncTasks.Wait()` should be specific to that use case? Maybe I'm wrong but I think that we should do that all the time before calling `Remove` effectively and closing the connection to other nodes.

runshenzhu commented:
@aaronlehmann That's my mistake. I just found that removing the time limit from context won't do any changes, because in the test `raftClient` also has a grpc time out limit. So, eventually, if fake clock didn't sync with real clock, `Leave` will time out, either from `context`, or from `grpc.invoke` inside the `raftClient.Leave`. When running the test this morning, I got the time out from grpc.

So there is still a time out problem. Test was passed just because it didn't hit the time out race. 

I fixed issues clock sync in test, and updated it to [my repo](https://github.com/runshenzhu/swarmkit/commits/raft-leave-join). 

while my test is crappy and throws meaningless errors, I think `n.asyncTasks.Wait()` approach is correct.

runshenzhu commented:
@abronan @aaronlehmann updated: use `n.asyncTasks.Wait()` as cleanup for all node leave cases.

aaronlehmann commented:
LGTM

abronan commented:
LGTM
